### PR TITLE
Load all files in `.a`

### PR DESCRIPTION
**Pull request by wargio**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

This will load all the `*.o` files embedded inside an `.a` file

I will keep this here, but i will postpone the completion of the PR

### CODE REVIEW

File: librz/core/cfile.c [hunk 1]
```diff
@@ -925,6 +925,7 @@ RZ_API bool rz_core_bin_load(RZ_NONNULL RzCore *r, RZ_NULLABLE const char *filen
 	bool is_io_load = false;
 	const char *cmd_load;
 	if (!cf) {
+		eprintf("NOPE\n");
```
Comment (XVilka) [inline librz/core/cfile.c]: ORLY?
Also `RZ_LOG_ERROR()`

File: librz/core/cfile.c [hunk 2]
```diff
@@ -1098,39 +1099,40 @@ RZ_API bool rz_core_bin_load(RZ_NONNULL RzCore *r, RZ_NULLABLE const char *filen
...
-			rz_core_bin_load(r, desc->name, loadaddr);
+
+		RzCoreFile *fh = core_file_new(r, desc->fd);
+		if (!fh) {
+			continue;
 		}
+		r->file = fh;
+		if (!rz_list_empty(r->files)) {
+			if (!rz_core_bin_load(r, desc->name, loadaddr)) {
+				eprintf("failed to load %s\n", desc->name);
```
Comment (XVilka) [inline librz/core/cfile.c]: `RZ_LOG_ERROR`

File: librz/core/cfile.c [hunk 3]
```diff
@@ -1173,10 +1176,12 @@ RZ_API RzCoreFile *rz_core_file_open(RzCore *r, const char *file, int flags, ut6
 		goto beach;
 	}
 
-	fh = core_file_new(r, fd->fd);
 	if (!fh) {
-		eprintf("core/file.c: rz_core_open failed to allocate RzCoreFile.\n");
-		goto beach;
+		fh = core_file_new(r, fd->fd);
+		if (!fh) {
+			eprintf("core/file.c: rz_core_open failed to allocate RzCoreFile.\n");
```
Comment (XVilka) [inline librz/core/cfile.c]: `RZ_LOG_ERROR`

File: librz/io/p/io_ar.c [hunk 4]
```diff
@@ -10,6 +10,46 @@ static bool rz_io_ar_plugin_open(RzIO *io, const char *file, bool many) {
...
 
+static RzList *rz_io_ar_open_many(RzIO *io, const char *file, int perm, int mode) {
+	const char *arname = strstr(file, "://");
+	if (!arname) {
+		return NULL;
+	}
+	arname += 3;
+
+	RzList *all = ar_open_all(arname, rz_sys_open_perms(perm));
+	if (!all) {
```
Comment (XVilka) [inline librz/io/p/io_ar.c:21]: Maybe add some error message here

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/core/cfile.c (hunk 1 above)
Comment (wargio) [reply librz/core/cfile.c]: i think i will switch that check to rz_return_val_if_fail

File: librz/core/cfile.c [hunk 5]
```diff
@@ -1082,55 +1082,54 @@ RZ_API bool rz_core_bin_load(RZ_NONNULL RzCore *r, RZ_NULLABLE const char *filen
...
 	return true;
 }
 
-RZ_API bool rz_core_file_open_many(RZ_NONNULL RzCore *r, RZ_NULLABLE const char *file, int perm, ut64 loadaddr) {
-	const bool openmany = rz_config_get_i(r->config, "file.openmany");
-	int opened_count = 0;
-	RzListIter *fd_iter, *iter2;
-	RzIODesc *desc;
-
+RZ_API RzCoreFile *rz_core_file_open_many(RZ_NONNULL RzCore *r, RZ_NULLABLE const char *file, int perm, ut64 loadaddr) {
```
Comment (XVilka) [inline librz/core/cfile.c]: Could you please add an ownership for the return value, also add a Doxygen?

File: librz/core/cfile.c [hunk 6]
```diff
@@ -1173,10 +1173,12 @@ RZ_API RzCoreFile *rz_core_file_open(RzCore *r, const char *file, int flags, ut6
 		goto beach;
 	}
 
-	fh = core_file_new(r, fd->fd);
 	if (!fh) {
-		eprintf("core/file.c: rz_core_open failed to allocate RzCoreFile.\n");
-		goto beach;
+		fh = core_file_new(r, fd->fd);
+		if (!fh) {
+			RZ_LOG_ERROR("core/file.c: rz_core_open failed to allocate RzCoreFile.\n");
```
Comment (XVilka) [inline librz/core/cfile.c]: Even the filename is wrong, probably better to remove it from the error message,

File: shlr/ar/ar.c [hunk 7]
```diff
@@ -212,16 +208,82 @@ static int ar_parse_header(RzArFp *arf, filetable *tbl, ut64 arsize) {
...
 #undef VERIFY_AR_NUM_FIELD
 
+/**
+ * \brief Open specific file withen a ar/lib file.
+ * \param arname the name of the .a file
+ * \return a list of files or NULL
+ *
+ * Open an ar/lib and returns all the object files inside it.
+ */
+RZ_API RzList *ar_open_all(const char *arname, int perm) {
```
Comment (XVilka) [inline shlr/ar/ar.c:218]: Can be the `arname` NULL?

File: shlr/ar/ar.c [hunk 8]
```diff
@@ -212,16 +208,82 @@ static int ar_parse_header(RzArFp *arf, filetable *tbl, ut64 arsize) {
...
  * \brief Open specific file withen a ar/lib file.
  * \param arname the name of the .a file
  * \param filename the name of file in the .a file that you wish to open
  * \return a handle of the internal filename or NULL
  *
- * Open an ar/lib file by name. If filename is NULL, then archive files will be
- * listed.
+ * Open an ar/lib file by name.
  */
 RZ_API RzArFp *ar_open_file(const char *arname, int perm, const char *filename) {
```
Comment (XVilka) [inline shlr/ar/ar.c:286]: Same here

**Review by XVilka — APPROVED** [review]
Overall looks good, please fix these minor things. Also, I wonder if it makes sense to move code from `shlr/ar/` to the `librz/io` or `librz/bin`, but that's for the next PR.

File: shlr/ar/ar.c (hunk 7 above)
Comment (wargio) [reply shlr/ar/ar.c:218]: it should not, but it will fail when opening the file.

**Review by XVilka — APPROVED** [review]

**Pull request merged**